### PR TITLE
TLD Open CL Support

### PR DESCRIPTION
**Pull request by Auron-X**
This pull request is based on/includes the pull request:
https://github.com/Itseez/opencv_contrib/commit/26f16e64382f5327426f417f9cdf201049cb6ea5
from June 12.
Updates:
1. Detector remake: to make all stages of detection independent using buffers in between. It is done for more clear algorithm work-flow and easier GPU implementations in future.
2. Added 1-st level of detector parallelization: OCL versions of Sr and Sc functions
3. Added 2-nd level of detector parallelization: Sr&Sc batch processing for scanning grid
4. Added OCL version of "integrateAdditional" function

### CODE REVIEW

**Comment by vpisarev** [comment]
I guess, this PR includes the previous one, but since we've merged that previous one, this one can not be merged anymore. Could you, please, update it?

File: modules/tracking/src/tldTracker.cpp [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    //best overlap around 92%
+
+    std::vector<Rect2d> candidates;
+    std::vector<double> candidatesRes;
+    bool trackerNeedsReInit = false;
+    for( int i = 0; i < 2; i++ )
+    {
+        Rect2d tmpCandid = boundingBox;
+        if( ( (i == 0) && !data->failedLastTime && trackerProxy->update(image, tmpCandid) ) ||
+			((i == 1) && (tldModel->detector->ocl_detect(imageForDetector, image_blurred, tmpCandid, detectorResults, tldModel->getMinSize()))))
```
Comment (lenlen) [inline modules/tracking/src/tldTracker.cpp]: Maybe here a condition if is needed for ocl_detect and detect methods, see this presentation https://docs.google.com/presentation/d/1qoa29N_B-s297-fp0-b3rBirvpzJQp8dCtllLQ4DVCY/present#slide=id.p26

Comment (Auron-X) [inline modules/tracking/src/tldTracker.cpp]: Yes, sure we should follow Transparent API rules. I will add this condition during cleaning step

**Comment by Auron-X** [comment]
Updated according to Vadim's comment

**Comment by lenlen** [comment]
@Auron-X there are some warnings in the build system, can you fix them?

**Comment by Auron-X** [comment]
Fixed

**Comment by vpisarev** [comment]
@Auron-X, thanks! I see a lot of commented off debugging code with dprintf's and such. While it's fine to retain the code at this stage, I hope you will not forget to clean the code in the end of the GSoC

**Comment by vpisarev** [comment]
:+1:

**Pull request merged**